### PR TITLE
Secure Cell overview

### PR DESCRIPTION
**Pull request by ilammy**
Start describing Secure Cell cryptosystem with an overview of its design, desired properties, and operation modes.

(I don't expect *everyone* on the reviewer list to chime in, just indicating that this PR might be interesting for you.)

### CODE REVIEW

File: themis/spec/secure-cell.md
Comment (vixentael) [inline themis/spec/secure-cell.md:207]:
```suggestion
If a passphrase is used, a [special _passphrase_ KDF](../../crypto-theory/cryptosystems/secure-cell.md#key-derivation-functions) is employed
```

**Review by vixentael — APPROVED** [review]
book-like spec! 👏

File: themis/spec/secure-cell.md
Comment (iamnotacake) [inline themis/spec/secure-cell.md:184]: > does not have any additional storage available

Sounds confusing (at least, for me). What about `does not require any additional storage`?
Or maybe `use` instead of `require`.

File: themis/spec/secure-cell.md
Comment (vixentael) [reply themis/spec/secure-cell.md:23]: I don't like this sentence either, I suggested a fix here
https://github.com/cossacklabs/product-docs/pull/30#discussion_r531084304

Comment (ilammy) [reply themis/spec/secure-cell.md:23]: AES-CTR in Context Imprint not only produces equally sized output, but the overall security of it is slightly lower than that of the Seal mode: the IV is completely deterministic, so if the user does not use unique contexts all the time, then Context Imprint may be susceptible to IV reuse attacks.

That's what I meant by the compromise: you sacrifice "secure by default" Seal mode for Context Imprint with no storage requirements but (slightly) lower baseline security, and again you can sacrifice some user-friendliness for additional security in Context Imprint mode.

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md:62]: Yeah, I don't really like this paragraph as it implies that Secure Cell is a communication protocol or something, with designated 'sender' and 'receiver'. What I want to convey is that Secure Cell uses shared secrets by design, contrasting it with Secure Message which lets parties keep their own secrets.

Any rewording suggestions?

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md:73]: In contexts like full-disk encryption it is a desired property to be able to start reading an encrypted file from an arbitrary location, without decrypting the entire file first (e.g., encrypted video files). However, Secure Cell does not support this use case, it was not designed for it, and will probably never support it.

(Both AES-GCM and AES-CTR are streaming ciphers at their core—so Secure Cell *could* support streaming decryption, if it were willing to ignore the authentication part of GCM. But it doesn't and won't.)

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md:184]: > Sounds confusing (at least, for me).

Point taken. I'll reword it as you have suggested.

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md:179]: > the resulted ciphertext has larger length that original plaintext

*Well ackshually,* the 'ciphertext' has exactly the same length in Seal mode too, since the authentication tag is not a part of the ciphertext (i.e., it's not required for decryption). Most of the extra fluff comes from the Secure Cell alone. If the users were expected to provide it all on decryption (like they are expected to do with the context for Context Imprint mode), it would have been possible to sorta-kinda preserve the length in Seal mode as well, with massive inconvenience pushed onto the users.

Okay, I'll give it a shot, but I'll probably reword it slightly...

File: themis/spec/secure-cell.md
Comment (storojs72) [reply themis/spec/secure-cell.md:62]: May be we can use encryptor/decryptor instead of sender/receiver?

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md]: Alrightly, how about this one that I've just pushed?

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md]: @storojs72, thanks for your valuable input! As I'm not a cryptographer myself in any way...

I've edited the paragraph to incorporate your suggestions. Could you please check it?

File: themis/spec/secure-cell.md
Comment (ilammy) [reply themis/spec/secure-cell.md]: Okay... how about the current version which just states things about symmetric encryption and shared secrets directly?

File: themis/spec/secure-cell.md
Comment (vixentael) [inline themis/spec/secure-cell.md:23]: 👍

File: themis/spec/secure-cell.md
Comment (vixentael) [inline themis/spec/secure-cell.md:62]: 👍

File: themis/spec/secure-cell.md
Comment (vixentael) [inline themis/spec/secure-cell.md:175]: 👍

**Review by vixentael — APPROVED** [review]
I like the changes very much.
I think we should wait for @storojs72 to approve attacks piece.

**Review by storojs72 — APPROVED** [review]
Looks good for me. Thank you @ilammy

File: themis/spec/secure-cell.md
Comment (storojs72) [reply themis/spec/secure-cell.md]: Looks good

File: themis/spec/secure-cell.md
Comment (storojs72) [reply themis/spec/secure-cell.md]: I think we can leave it up to you, since it is a wording

**Review by iamnotacake — APPROVED** [review]

**Comment by ilammy** [comment]
Alrighty... It seems we're done here. Thanks everyone for review! That's really helpful ❤️

I guess there will be another round once the page is complete and it gets to staging. I'll see to it.

**Pull request merged**
